File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
#2206 벽부수고 이동하기
from collections import deque
def road_find(maps):
    queue = deque([[0,0,0]])
    visit = [[[0, 0] for _ in range(m)] for _ in range(n)]
    visit[0][0][0] = 1
    dx, dy = [1,-1,0,0],[0,0,1,-1]
    while(queue):
            x, y, c = queue.popleft()
            if(x == n-1 and y == m-1):
                return visit[x][y][c]
            for i in range(4):
                a, b = x+dx[i], y+dy[i]
                if(0<=a<n and 0<=b<m):
                    if(maps[a][b] == 0 and visit[a][b][c] == 0):
                        visit[a][b][c] = visit[x][y][c] + 1
                        queue.append([a,b,c])
                    if(maps[a][b] == 1 and c == 0):
                        visit[a][b][c+1] = visit[x][y][c] +1
                        queue.append([a,b,c+1])
    return -1
        
n, m = map(int, input().split())
maps = []
for _ in range(n):
    maps.append(list(map(int, input())))
    
print(road_find(maps))

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <iostream> // 표준 입출력
#include <vector> // 벡터 (STACK)
#include <queue> // 우선순위 큐
#include <algorithm> // 정렬 
#include <string>  // 문자열
#include <numeric>
#include <math.h>
#include<cstring>

using namespace std;

int dx[4] = { 0, 0, 1, -1 };
int dy[4] = { 1, -1, 0 ,0 };

string maps[1001];
int visit[1001][1001][2];

int solution(int n, int m) {
	/*
		2206 벽 부수고 이동하기
		- (1, 1) -> (n, n) 이동
		- 이동하는 도중 한개의 벽을 부수고 이동 가능
	*/
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			visit[i][j][0] = 9000000;
			visit[i][j][1] = 9000000;
		}
	}

	queue<vector<int>> q;
	q.push({ 0, 0, 1, 0});
	while (!q.empty()) {
		auto now = q.front(); // x, y, 이동 횟수, 벽 부셧나요
		int x = now[0]; int y = now[1]; int move = now[2]; int wall = now[3];
		q.pop();
		if (x== n - 1 && y == m - 1) {
			return move;
		}
		for (int k = 0; k < 4; k++) {
			int nx = x + dx[k]; 
			int ny = y + dy[k];
			if (0 <= nx && nx <= n - 1 && 0 <= ny && ny <= m - 1) {
				if (maps[nx][ny] == '0' && visit[nx][ny][wall] > move + 1) {
					visit[nx][ny][wall] = move + 1;
					q.push({ nx, ny, move + 1, wall });
				}
				if (maps[nx][ny] == '1' && wall == 0) {
					if (visit[nx][ny][1] > move + 1) {
						visit[nx][ny][1] = move + 1;
						q.push({ nx, ny, move + 1, 1 });
					}
				}
			}
		}
	}
	return -1;
}

int main() {
	int n, m; cin >> n >> m;
	for (int i = 0; i < n; i++) {
		cin >> maps[i];
	}
	cout << solution(n , m);
}